### ISSUE TITLE
`ignore_errors: {{ ansible_check_mode }}` is always true

### ISSUE THREAD

**Issue by cdelgehier** [ansible/ansible#56345]
##### SUMMARY
When having a task which uses `ignore_errors: {{ ansible_check_mode }}` and the task results in an exception the error is always ignored. 
This is especially bad when running in non-check mode.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
task_executor

##### ANSIBLE VERSION

```
ansible 2.7.10
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/cedric/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/cedric/test/venv/lib/python2.7/site-packages/ansible
  executable location = /home/cedric/test/venv/bin/ansible
  python version = 2.7.5 (default, Apr 11 2018, 07:36:10) [GCC 4.8.5 20150623 (Red Hat 4.8.5-28)]
```

##### CONFIGURATION

```paste below
nothing
```

##### STEPS TO REPRODUCE

```yaml
- name: bug on ignore_errors and ansible_check_mode
  hosts: localhost
  gather_facts: false

  tasks:
    - debug:
        var: ansible_check_mode

    - debug:
        msg: "foo = {{ foo }}"
      when: foo
      ignore_errors: "{{ ansible_check_mode }}"
```

##### EXPECTED RESULTS

Ignore errors only in the check mode, as indicated in the documentation https://docs.ansible.com/ansible/latest/user_guide/playbooks_checkmode.html?information-about-check-mode-in-variables#information-about-check-mode-in-variables

##### ACTUAL RESULTS

```paste below
(venv) cedric@jacksbastion ~/test $ ansible-playbook test.yml -vv
ansible-playbook 2.7.10
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/cedric/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/cedric/test/venv/lib/python2.7/site-packages/ansible
  executable location = /home/cedric/test/venv/bin/ansible-playbook
  python version = 2.7.5 (default, Apr 11 2018, 07:36:10) [GCC 4.8.5 20150623 (Red Hat 4.8.5-28)]
Using /etc/ansible/ansible.cfg as config file
/etc/ansible/hosts did not meet host_list requirements, check plugin documentation if this is unexpected
/etc/ansible/hosts did not meet script requirements, check plugin documentation if this is unexpected
 [WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'


PLAYBOOK: test.yml *****************************************************************************************************************************************************************************************
1 plays in test.yml

PLAY [bug on ignore_errors and ansible_check_mode] *********************************************************************************************************************************************************
META: ran handlers

TASK [debug] ***********************************************************************************************************************************************************************************************
task path: /home/cedric/test/test.yml:6
ok: [localhost] => {
    "ansible_check_mode": false
}

TASK [debug] ***********************************************************************************************************************************************************************************************
task path: /home/cedric/test/test.yml:9
fatal: [localhost]: FAILED! => {"msg": "The conditional check 'foo' failed. The error was: error while evaluating conditional (foo): 'foo' is undefined\n\nThe error appears to have been in '/home/cedric/test/test.yml': line 9, column 7, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n\n    - debug:\n      ^ here\n"}
...ignoring
META: ran handlers
META: ran handlers

PLAY RECAP *************************************************************************************************************************************************************************************************
localhost                  : ok=2    changed=0    unreachable=0    failed=0
```
The `foo` variable is undefined but the play ignores the error `...ignoring`

**Comment by ansibot** [comment]
Files identified in the description:
* [lib/ansible/executor/task_executor.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/executor/task_executor.py)

If these files are inaccurate, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by sivel** [comment]
The difference here, between the documentation, and the example in this issue, is that your issue is caused by an undefined variable.

The undefined variable is encountered before we can template `ignore_errors`, and since `ignore_errors` is a string at that point, that is considered truthy in Python.

If your example was instead:

```
      fail:
      ignore_errors: "{{ ansible_check_mode }}"
```

It would work as expected.

**Comment by cdelgehier** [comment]
OK 
```
- name: bug on ignore_errors and ansible_check_mode
  hosts: localhost
  gather_facts: false

  tasks:
    - fail:
        msg: "foo = bar"
      ignore_errors: "{{ ansible_check_mode }}"
```
```
(venv) cedric@jacksbastion ~/test $ ansible-playbook test.yml -vv
ansible-playbook 2.7.10
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/cedric/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/cedric/test/venv/lib/python2.7/site-packages/ansible
  executable location = /home/cedric/test/venv/bin/ansible-playbook
  python version = 2.7.5 (default, Apr 11 2018, 07:36:10) [GCC 4.8.5 20150623 (Red Hat 4.8.5-28)]
Using /etc/ansible/ansible.cfg as config file
/etc/ansible/hosts did not meet host_list requirements, check plugin documentation if this is unexpected
/etc/ansible/hosts did not meet script requirements, check plugin documentation if this is unexpected
 [WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit
localhost does not match 'all'


PLAYBOOK: test.yml **********************************************************************************
1 plays in test.yml

PLAY [bug on ignore_errors and ansible_check_mode] **************************************************
META: ran handlers

TASK [fail] *****************************************************************************************
task path: /home/cedric/test/test.yml:6
fatal: [localhost]: FAILED! => {"changed": false, "msg": "foo = bar"}
	to retry, use: --limit @/home/cedric/test/test.retry

PLAY RECAP ******************************************************************************************
localhost                  : ok=0    changed=0    unreachable=0    failed=1
```
```
(venv) cedric@jacksbastion ~/test $ ansible-playbook test.yml -vv --check
ansible-playbook 2.7.10
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/cedric/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/cedric/test/venv/lib/python2.7/site-packages/ansible
  executable location = /home/cedric/test/venv/bin/ansible-playbook
  python version = 2.7.5 (default, Apr 11 2018, 07:36:10) [GCC 4.8.5 20150623 (Red Hat 4.8.5-28)]
Using /etc/ansible/ansible.cfg as config file
/etc/ansible/hosts did not meet host_list requirements, check plugin documentation if this is unexpected
/etc/ansible/hosts did not meet script requirements, check plugin documentation if this is unexpected
 [WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit
localhost does not match 'all'


PLAYBOOK: test.yml **********************************************************************************
1 plays in test.yml

PLAY [bug on ignore_errors and ansible_check_mode] **************************************************
META: ran handlers

TASK [fail] *****************************************************************************************
task path: /home/cedric/test/test.yml:6
fatal: [localhost]: FAILED! => {"changed": false, "msg": "foo = bar"}
...ignoring
META: ran handlers
META: ran handlers

PLAY RECAP ******************************************************************************************
localhost                  : ok=1    changed=0    unreachable=0    failed=0
```

**Comment by sivel** [comment]
We have discussed this in the recent core IRC meeting, and have decided that undefined vars should not be task errors, but should instead become a parser error that would cause the playbook to abort.

This will likely be a change in behavior, so it will also need a porting guide entry describing this change.

In short, `ignore_errors` should not be able to impact the outcome of an undefined variable.

**Comment by ssbarnea** [comment]
It seems that I was hit but the same weird behavior where code below failed to work
```
      ignore_errors: ansible_distribution == 'RedHat'
```

```
      ignore_errors: "{{ ansible_distribution == 'RedHat' }}"
```

Even using jijnja2 notation ended up in the same silent failure, evaluated as `false`.

**Comment by sivel** [comment]
I've looked into this a little bit, and I believe the solution is going to be a little difficult.

The problem is that the task, is run through `post_validate` in `TaskExecutor` which happens in a fork. It does raise `AnsibleParserError`, but the only way to relay that back to the controller is via a module failure which is why we see the current behavior.

With recent changes we now have a `get_validated_value` function that will perform the `isa` type checking.  I see 2 options:

1. Update the `_generic_g*` functions to call `get_validated_value` (I experimented briefly with this, but had issues)
2. Individually use `get_validated_value` for certain things such as `original_task.ignore_errors` in `StrategyBase`.

In addition to the above, a few changes would need to be made to `get_validated_value` to allow this, such as, if `attribute` is `None`, do the lookup using the `name` in `self._valid_attrs` for the caller since not all callers would have access to that.

Also, a weird one is `isa='class'` which seems to be the only one to require `Templar`.

I am not proceeding with working on this issue at this time.

**Comment by LuckySB** [comment]
follow up https://github.com/ansible/ansible/issues/61025

**Comment by andras-tim** [comment]
For the record, this still exists in **ansible [core 2.12.2]** and makes a lot of pain for me too.

E.g. I have written a bad comparison operator (`===`), but independent of the `--check` this will be always ignored.

``` yaml
- name: '[{{ username }}] Ensure user shell is set properly'
  when: |-
    has_access and user_info.shell === '/bin/sh'
  user:
    name: "{{ username }}"
    shell: '/bin/bash'
  ignore_errors: "{{ ansible_check_mode }}"
```

... and I have no workaround.

**Comment by jbyibm** [comment]
I have the same situation. See https://github.com/ansible/ansible/issues/79460. 
I read this issue and if I understand well, there is no fix or workaround?

**Comment by cdelgehier** [comment]
yep

**Comment by jbyibm** [comment]
For me Ansible controller should handle this and stop on this case.

**Comment by DE110283** [comment]
Please fix this - or please update the official ansible documentation,

**Comment by Conhokis** [comment]
Still present in 2.13.11

**Comment by ansibot** [comment]
@cdelgehier: Thank you for taking the time to open this issue. In order for the community to handle your issue effectively, we need the appropriate template to be filled in.

Here are the items we could not find in your description:
- OS / Environment

As such we are closing this issue. Please open a [new issue](https://github.com/ansible/ansible/issues/new/choose) and choose the appropriate template.

[click here for bot help](https://github.com/ansible/ansibotmini#ansibotmini)
